### PR TITLE
Per file settings

### PR DESCRIPTION
**Pull request by harshad1**
This PR adds the following toggles in the document menu

1. Wrap state
2. Enable / Disable highlighting

Additionally, this PR adds the ability to set indent per file by long pressing the indent actions, and the ability to remember the last mode (view vs edit)

Also included is some code to reduce the errors caused by AccessibilityEvents in longer files. Accessibility is disabled in files > 10k characters. AFAIK this is an android limitation. Above 10k characters, accessibility can cause crashes and data loss.

### CODE REVIEW

**Comment by harshad1** [comment]
@gsantner As discussed, please let me know which (if any) other options should have toggles in the document menu

File: app/src/main/res/menu/document__edit__menu.xml [hunk 1]
```diff
@@ -79,22 +80,39 @@
...
         android:title="@string/search"
         android:icon="@drawable/ic_search_black_24dp"
         app:showAsAction="always"
         app:actionViewClass="android.widget.SearchView" />
+
     <item
         android:id="@+id/action_save"
         android:icon="@drawable/ic_save_black_24dp"
         android:title="@string/save"
         app:showAsAction="never" />
```
Comment (gsantner) [inline app/src/main/res/menu/document__edit__menu.xml:101]: as first step, please wrap the new 2 into a common submenu

**Comment by gsantner** [comment]
I'm not totally sure if it's the right path with checkbox menu options (as we are limited to exactly that with mrnu then)... we will see

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 2]
```diff
@@ -150,22 +145,44 @@ private void cancelUpdate() {
...
+        highlightWithoutChange();
+    }
+
+    // This allows us to disable sending of accessibility events when highlighting is being
+    // performed. Sending of many accessibility events can cause performance issues and crashes.
+    // Blocking accessibility during _highlighting_ should have no impact on users.
+    @Override
+    public void sendAccessibilityEventUnchecked(AccessibilityEvent event) {
+        if (_accessibilityEnabled && getText().length() < 10000) {
+            super.sendAccessibilityEventUnchecked(event);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:157]: @gsantner Added the fix for the accessibility / crash issue to this PR as there were some conflicting changes. I will close the other PR. This fix is a little better, but we still need to determine the correct max length and what the impact is on accessibility users.

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java [hunk 3]
```diff
@@ -120,6 +128,8 @@ public static DocumentEditFragment newInstance(File path, boolean pathIsFolder,
     private boolean _firstFileLoad = true;
 
     public DocumentEditFragment() {
+        super();
+        _appSettings = AppSettings.get();
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java:132]: preferably, please use one of the init / onCreate method overrides to init with fragment context .. avoid AppSettings.get() if possible

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java (hunk 2 above)
Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:157]: as far I can see accessbility is only disabled during highlighting runs, right? if so, should be fine

Comment (harshad1) [reply app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:169]: Am still trying to figure out the minimal impact fix here. New version does 2 thing:
1. Block accessibility during highlighting
2. If the text in the accessibility message is too large, then cut it down

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java (hunk 3 above)
Comment (harshad1) [reply app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java]: Fixed

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java (hunk 2 above)
Comment (harshad1) [reply app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:169]: This seems to be working so far. If we can do this properly, then there should be no impact on accessibility users while solving the crash issue

**Comment by harshad1** [comment]
> @gsantner As discussed, please let me know which (if any) other options should have toggles in the document menu

**Comment by harshad1** [comment]
@gsantner Added the long press indent button to change setting here (seemed logical)

Accessibility is currently disabled above 10k characters. Should we add an experimental setting to allow this to be overwritten 

-[ ] Enable accessibility for large files
Warning: may cause crashes and data loss

**Comment by harshad1** [comment]
This is ready to go.

If you have suggestions of what else we want to add to File options menu I can add them here.

Or we can always make a follow up or.

**Comment by harshad1** [comment]
@gsantner I have had absolutely no crashes or performance issues since implementing the 10000 char limit for accessibility here.

**Comment by harshad1** [comment]
Do you think adding a per file 'open in view mode' makes sense?

**Comment by gsantner** [comment]
Storing last state for current file, and apply again when open next time is a better option a would say

**Comment by harshad1** [comment]
> Storing last state for current file, and apply again when open next time is a better option

Implemented and pushed. I have removed the global setting as it doesn't make sense any more.

File: app/src/main/java/net/gsantner/opoc/util/StringUtils.java [hunk 4]
```diff
@@ -245,4 +248,14 @@ public static String unescapeString(final String input) {
 
         return builder.toString();
     }
+
+    public static String getPath(final Document document) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/opoc/util/StringUtils.java:252]: please don't add markor.Document to StringUtils if not required, which is not here.  
For example, have it in DocumentIO or at the other getPath() place

**Comment by gsantner** [comment]
Those people that ask specifically for "how to always open directly in view mode" are those who usually check the "Always view mode" settings option. Thus just always return true when it is checked in settings. For false case we return the (NEW) last state.

pushed that small change

**Comment by harshad1** [comment]
> pushed that small change

Nice approach,

I have moved getPath to Document (still as a static function)

**Comment by gsantner** [comment]
`it's ok to be static` :D.

I spent like 1h-2h now to go through all the PRs and issues..did quite some things. Let me know if you have some feedback somewhere.

What do you think, still something todo in this PR, something you want to test?

**Comment by harshad1** [comment]
> I spent like 1h-2h now to go through all the PRs and issues..did quite some things. Let me know if you have some feedback somewhere.

I see all the emails :D.

> What do you think, still something todo in this PR, something you want to test?

Want to spend a day testing the more recent changes. I don't want to add any more to this PR.

**Comment by harshad1** [comment]
@gsantner this is ready to go

**Comment by gsantner** [comment]
thanks much, merged! 

If you have time, can you please help review/test test the Zim PR? The most important point is that it not break Markdown and other formats (in regards of the ListHandler, AutoFormat, Highlighter, TextActions etc).  
Due you refactored a lot there recently you know that best atm, so if you can help to check that would be cool.  
However, don't look at Zim format, focus is on "not break something else except Zim". Zim is the first/initial version and it's ok when it's not 100% done right away.

**Comment by harshad1** [comment]
> If you have time, can you please help review/test test the Zim PR? The most important point is that it not break Markdown and other formats (in regards of the ListHandler, AutoFormat, Highlighter, TextActions etc).

Will do.

**Pull request merged**